File: src/main/java/_sortByNumber/_101_200/_142/Solution.java
package _sortByNumber._101_200._142;

class ListNode {

    int val;
    ListNode next;
    ListNode(int x) {val = x; next = null;}
}

public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) return head;
        ListNode p1 = head, p2 = head;
        boolean hasCycle = false;
        while (p2.next != null && p2.next.next != null) {
            p1 = p1.next;
            p2 = p2.next.next;
            if (p1 == p2) {
                hasCycle = true;
                break;
            }
        }
        if (hasCycle) {
            ListNode p3 = head;
            while (p3 != p1) {
                p3 = p3.next;
                p1 = p1.next;
            }
            return p3;
        } else
            return null;

    }
}

File: src/main/java/_sortByNumber/_101_200/_142/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
from typing import Optional

'''给定一个链表的头节点  head ，返回链表开始入环的第一个节点。 如果链表无环，则返回 null。'''
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast, slow = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow ==  fast:
                ans = head
                while ans != slow:
                    ans = ans.next
                    slow = slow.next
                return ans